### PR TITLE
Add caching of Docker images for CI

### PR DESCRIPTION
**Pull request by artemnikitin**
At the moment, only Docker image for CI was added. In the future we can add more images, like from https://github.com/elastic/cloud-on-k8s/issues/1096

### CODE REVIEW

**Comment by david-kow** [comment]
How this works exactly? Is this executed once to prepare the image which is reused during many ci runs?

**Comment by artemnikitin** [comment]
@david-kow no idea. In linked `infra` issue https://github.com/elastic/infra/issues/13910 there is a quick description how to setup it

**Comment by david-kow** [comment]
Thanks. Found https://github.com/elastic/infra/blob/master/docs/packer2.0/caching.md there (tldr: image rebuilt daily, reused for multiple ci runs).

File: .ci/packer_cache.sh [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+# This script takes responsibility to create a minikube cluster. It has all
+# of the necessary default settings so that no environment variable has to
+# be specified.
+
+declare -a docker_images=("docker.elastic.co/eck/eck-ci:57e169ff162c2ea351a1956e1ad355b9")
+
+# Pull all the required docker images
+for image in "${docker_images[@]}"
+do
+  docker pull $image
```
Comment (jmlrt) [inline .ci/packer_cache.sh]:
```suggestion
  docker pull "$image"
```

File: .ci/packer_cache.sh [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+# Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+# or more contributor license agreements. Licensed under the Elastic License;
+# you may not use this file except in compliance with the Elastic License.
+
+# This script takes responsibility to create a minikube cluster. It has all
+# of the necessary default settings so that no environment variable has to
+# be specified.
+
+declare -a docker_images=("docker.elastic.co/eck/eck-ci:57e169ff162c2ea351a1956e1ad355b9")
```
Comment (sebgl) [inline .ci/packer_cache.sh]: What's the update process? Do we need to manually change the tag here?
Why not computing the tag value automatically here?

File: .ci/packer_cache.sh [hunk 3]
```diff
@@ -0,0 +1,17 @@
+#! /usr/bin/env bash
+
+# Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+# or more contributor license agreements. Licensed under the Elastic License;
+# you may not use this file except in compliance with the Elastic License.
+
+# This script takes responsibility to create a minikube cluster. It has all
+# of the necessary default settings so that no environment variable has to
+# be specified.
```
Comment (sebgl) [inline .ci/packer_cache.sh]: Comments looks wrong.

File: .ci/packer_cache.sh [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+
```
Comment (jmlrt) [inline .ci/packer_cache.sh]: We recommend adding `set -eou pipefail` in the beginning of the script to fail if we encounter an error during packer cache

File: .ci/packer_cache.sh (hunk 3 above)
Comment (artemnikitin) [reply .ci/packer_cache.sh]: Fixed

File: .ci/packer_cache.sh [hunk 5]
```diff
@@ -0,0 +1,17 @@
+#! /usr/bin/env bash
+
+# Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+# or more contributor license agreements. Licensed under the Elastic License;
+# you may not use this file except in compliance with the Elastic License.
+
+# This script takes responsibility to create a minikube cluster. It has all
```
[the remark this answers is not in the document]
Comment (artemnikitin) [reply .ci/packer_cache.sh]: Fixed

File: .ci/packer_cache.sh (hunk 4 above)
Comment (artemnikitin) [reply .ci/packer_cache.sh]: Fixed

File: .ci/packer_cache.sh (hunk 1 above)
Comment (artemnikitin) [reply .ci/packer_cache.sh]: Fixed

File: .ci/packer_cache.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (artemnikitin) [reply .ci/packer_cache.sh]: Quite rare. Probably much rare than once per week.

Comment (artemnikitin) [reply .ci/packer_cache.sh]: @jmlrt When `packer_cache.sh` is executed do we have an access to the git repo content?

Comment (jmlrt) [reply .ci/packer_cache.sh]: Yes, the full repository is cloned the `packer_cache.sh` is executed from inside the repository directory ([ansible/roles/packer_cache/tasks/linux/packer_cache.yml#L26-L39](https://github.com/elastic/infra/blob/master/ansible/roles/packer_cache/tasks/linux/packer_cache.yml#L26-L39))

Comment (artemnikitin) [reply .ci/packer_cache.sh]: @jmlrt On the machine where it will be executed do we have `make`, `md5sum` and `awk`?

Comment (jmlrt) [reply .ci/packer_cache.sh]: I can confirm you that they are already installed when `packer_cache.sh` script is run.

File: .ci/packer_cache.sh [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+# This script used to "bake" Docker images into base image for Jenkins nodes.
+
+set -eou pipefail
+
+DOCKER_CI_IMAGE=$(cd ../build/ci/ && make show-image)
+
+declare -a docker_images=("$DOCKER_CI_IMAGE")
+
+# Pull all the required docker images
+for image in "${docker_images[@]}"
```
Comment (sebgl) [inline .ci/packer_cache.sh]: not sure we need to iterate on an array of 1 element 🤷‍♂

Comment (sebgl) [reply .ci/packer_cache.sh]: what if you cannot pull the image because it's not built yet? will it make the CI process fail?

Comment (jmlrt) [reply .ci/packer_cache.sh]: If you cannot pull the docker image, then the Packer image build for ECK workers will fail but we can still relaunch the build or wait next build the next day. Meanwhile ECK workers can still use the Packer image from the day before.

File: .ci/packer_cache.sh (hunk 2 above)
Comment (artemnikitin) [reply .ci/packer_cache.sh]: Great news :)

Comment (artemnikitin) [reply .ci/packer_cache.sh]: Fixed. It will use makefile to get tag for image, same as on CI

File: .ci/packer_cache.sh (hunk 6 above)
Comment (artemnikitin) [reply .ci/packer_cache.sh]: > not sure we need to iterate on an array of 1 element 🤷‍♂

In the future we will add more images, like for `kind`

Comment (artemnikitin) [reply .ci/packer_cache.sh]: > will it make the CI process fail?

It will not affect our builds on CI

**Pull request merged**